packages/react-components/react-utilities/scripts/expand-jsx-intrinsic-elements.ts

```diff
@@ -0,0 +1,297 @@
+import * as ts from 'typescript';
+import * as fs from 'node:fs';
+import * as path from 'node:path';
+import { execSync } from 'node:child_process';
+import { parseArgs } from 'node:util';
+
+// Run the transformation
+if (require.main === module) {
+  const { omitElements, reactTypesPath, targetFile } = processArgs();
+  main({
+    omitElements,
+    reactTypesPath,
+    targetFile,
+  }).catch(console.error);
+}
+
+export { main };
+
+/**
+ * Advanced TypeScript transform script that dynamically extracts the actual keys
+ * from JSX.IntrinsicElements by analyzing the React type definitions
+ */
+async function main(options: { targetFile: string; reactTypesPath: string; omitElements: string[] }) {
+  const { reactTypesPath, targetFile, omitElements } = options;
+
+  console.log('🔄 Extracting JSX intrinsic element keys from React types...');
+  const elementKeys = extractJSXKeysFromReactTypes({ reactTypesPath });
+  console.log(`📋 Found ${elementKeys.length} JSX intrinsic elements`);
+
+  // Filter out omitted elements
+  const filteredElementKeys = elementKeys.filter(key => !omitElements.includes(key));
+
+  if (omitElements.length > 0) {
+    const omittedCount = elementKeys.length - filteredElementKeys.length;
+    console.log(`🚫 Omitted ${omittedCount} elements: ${omitElements.join(', ')}`);
+    console.log(`✅ Remaining ${filteredElementKeys.length} elements after filtering`);
+  }
+
+  const unionString = filteredElementKeys.map(key => `'${key}'`).join(' | ');
+  const JSXIntrinsicElementKeysTypeStatement = `
+  /**
+  * Unwrapped type for 'keyof JSX.IntrinsicElement'.
+  */
+  export type JSXIntrinsicElementKeys = ${unionString};
+  \n`;
+
+  return generateFile({
+    targetFile,
+    types: [JSXIntrinsicElementKeysTypeStatement],
+  });
+}
+
+function processArgs() {
+  const { values } = parseArgs({
+    args: process.argv.slice(2),
+    options: {
+      'react-types-path': {
+        type: 'string',
+        short: 'r',
+        description: 'Path to the React types directory',
+      },
+      'target-file': {
+        type: 'string',
+        short: 't',
+        description: 'Output file path',
+      },
+      'omit-elements': {
+        type: 'string',
+        short: 'o',
+        description: 'Comma-separated list of elements to exclude from the generated union',
+      },
+      help: {
+        type: 'boolean',
+        short: 'h',
+        description: 'Show help message',
+      },
+    },
+    allowPositionals: false,
+  });
+
+  // Show help if requested
+  if (values.help) {
+    console.log(`
+Usage: node scripts/expand-jsx-intrinsic-elements.ts [options]
+
+Options:
+  -r, --react-types-path <path>    Path to the React types directory
+  -t, --target-file <path>         Output file path (default: ../src/utils/generated-types.ts)
+  -o, --omit-elements <elements>   Comma-separated list of elements to exclude to make JSXIntrinsicElements union backwards compatible (default: set,mpath,center,search)
```

Comment:
to be discussed

**The problem**
old major react versions don't contain the new expansion of `React.JSX.IntrinsicElements`

and because the generic constraint of `React.ComponentType<Element extends keyof React.JSX.IntrinsicElement>` , which we use as alias for `export type JSXIntrinsicElement<Element extends JSXIntrinsicElementKeys> = React.ComponentProps<Element>;` will explode if we obtained the union from rect types v18 or v19 and user of r17 would use our library.